CIP-0029/phase-1-monetary-scripts.cddl

```diff
@@ -0,0 +1,29 @@
+script =
+  [  script_key
+  // script_all
+  // script_any
+  // script_n_of_k
+  // script_after
+  // script_before
+  ]
+
+script_key =
+  (0, keyhash)
+
+script_all =
+  (1, [ * script ])
+
+script_any =
+  (2, [ * script ])
+
+script_n_of_k =
+  (3, n: uint, [ * script ])
+
+script_after =
+  (4, uint)
+
+script_before =
+  (5, uint)
```

Comment:
These are backwards compared to the Allegra & Mary CDDL spec.

Reply:
What do you mean with _backward_? I've tried to keep the terminology the same as the one used in the cardano-api. The cddl calls them "invalid before" and "invalid hereafter"

Reply:
The numbers are backwards compared to the Allegra & Mary CDDL spec.
```invalid_before = (4, uint)
invalid_hereafter = (5, uint)
```

Reply:
I think the number are correct, but the terminology is confusing. While the CDDL on cardano-ledger-specs speaks about *invalid* before for tag 4 (resp. invalid after for tag 5), this CIP CDDL (which uses the same terminology as the cardano-api) speaks about *valid* after for tag 4 (resp. valid before for tag 5).

From `cardano-api`:

```hs
  toJSON (RequireTimeAfter _ slot) =
    object [ "type" .= String "after"
           , "slot" .= slot
           ]
```

and

```hs
    go (RequireTimeAfter  _ t) = Timelock.RequireTimeStart  t
```

and from `cardano-ledger-specs`

```hs
pattern RequireTimeStart mslot <-
  TimelockConstr (Memo (TimeStart mslot) _)
```

```hs
encRaw (TimeStart m) = Sum TimeStart 4 !> To m
```

Thus, the JSON `"after"` is indeed encoded with tag `4` in CDDL, and `"after"` really means _valid after_ in this context. 

I agree that the discrepancy is unfortunate, and in my opinion, it should be fixed in the CDDL (since it's only a terminology change, whereas changing it in the JSON interface would be backward incompatible), and that's why I chose to stick with the cardano-api's JSON terminology in this proposal as well.